### ISSUE TITLE
Tone Mapping preview inaccurate

### ISSUE THREAD

**Issue by Beep6581** [Beep6581/RawTherapee#2233]
Originally reported on Google Code with ID 2249

```
Here is a patch that tries to provide a solution to bug spotted here.
http://www.rawtherapee.com/forum/viewtopic.php?f=3&t=5290

As for CbDL we cannot cut the pixels ...
But I reduced the action of "edgestopping" according to skip.

Maybe this correction is sufficient ?

:)

```

Reported by `jdesmis` on 2014-02-15 13:17:15

<hr>
- _Attachment: [issue2250-1.patch](https://storage.googleapis.com/google-code-attachments/rawtherapee/issue-2249/comment-0/issue2250-1.patch)_

**Comment by Hombre57** [comment]
@Desmis @heckflosse @Beep6581 @iliasg

I've created an [`epd-preview-accuracy`](https://github.com/Beep6581/RawTherapee/compare/epd-preview-accuracy) branch where I've made a small change (too small to be true, but it seem to partly solve the difference).

However there's still a problem of amplitude of the effect. [Here](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/EdgePreservingDecomposition.cc#L744) we see that a gradient is computed based on a four pixel square, which will give different result depending on the zoom scale I presume. I'll let those who knows look at the code, if the preview can be even closer to the output image. Otherwise we'll have to live with it.

**Comment by Beep6581** [comment]
Great! Will test it soon.

Heads up to testers: use Edge Stopping=1.4 or more with Scale=1 or more.

**Comment by Beep6581** [comment]
Tested at a zoom of 25%, alt+tabbing between patched and unpatched RT and a image saved at the same size as the preview (973px wide).

I am thrilled to see that with Edge Stopping (ES) at 1.4 or less, the other values now just about perfectly match the saved image!

Tested with Edge Stopping (ES) at 1.4, Scale (S) at 1 and RI at 0, the preview is just about identical to the saved image. Still almost identical at ES=1.4 S=1 RI=10.

However setting ES more than 1.4 results in a wildly different preview to saved image, as it was before the patch.

No differences between patched and unpatched in saved files (good).

Some time ago I changed the defaults to match the saved image better, but that meant making them weak to the point where they didn't do much. Thanks to this improvement, I think we should set the default values back to:
ES=1.4
S=1
RI=0

```diff
diff --git a/rtengine/procparams.cc b/rtengine/procparams.cc
index 94694fc0..01e645c3 100644
--- a/rtengine/procparams.cc
+++ b/rtengine/procparams.cc
@@ -1185,8 +1185,8 @@ void ProcParams::setDefaults ()
     epd.enabled = false;
     epd.strength = 0.5;
     epd.gamma = 1.0;
-    epd.edgeStopping = 0.5;
-    epd.scale = 0.1;
+    epd.edgeStopping = 1.4;
+    epd.scale = 1.0;
     epd.reweightingIterates = 0;
 
     sh.enabled       = false;
```

**Comment by Hombre57** [comment]
Thanks for testing the patch @Beep6581 & @heckflosse. Branch merged into dev as discussed on IRC. Branch deleted, but this issue remains opened, so preview accuracy could be even more closer.

You can e.g. test with preview = 10%, S=2 and RI=0 and output with RI=2 (same other sliders) -> the preview strangely match the output !?

**Comment by Desmis** [comment]
@Hombre57 

RT devient comme Lucky Luke, plus rapide que son ombre. même pas le temps de respirer :)

Je viens de tester ton patch, il marche au poil.

Merci
Jacques

**Comment by agriggio** [comment]
@Beep6581 it seems your new proposed defaults are very aggressive compared to what we have now. I frequently use the tone mapping tool, and though I agree that the current defaults are generally too low, I like the fact that it acts in a subtle way...

**Comment by Beep6581** [comment]
@agriggio these defaults are not new, they've been like that for most of that tool's life, until I recently changed them because the preview looked very different to the saved image and people were complaining.
I remember either the original developer Ben or maybe Emil saying that ES is set to 1.4 by default for some mathematical reason. Now that the preview matches, we could use that again. Strength was originally at 0.25 though.

**Comment by agriggio** [comment]
ok, thanks for the clarification

**Comment by Hombre57** [comment]
@Beep6581 @heckflosse @Desmis I'll commit a patch in the coming hour, please don't release 5.2 in the mean time. I've found a location in the code that need the same correction than my previous patch, when "Use ToneMapping for EPD" is enabled (the preview was inaccurate).

**Comment by Hombre57** [comment]
Patch committed. Still not ideal regarding the overall contrast, but way better !

@Beep6581 : here is a pp3 for the HDR image I sent you that will demonstrate the bug. Try before and after patching.

[IMGP1198-1202-2.pp3.txt](https://github.com/Beep6581/RawTherapee/files/1168220/IMGP1198-1202-2.pp3.txt)

**Comment by Beep6581** [comment]
Yes, it seems a bit better in CIECAM02 mode, and it doesn't seem to cause any regression.

**Comment by Hombre57** [comment]
Forgot to explain that preview is better with `Reweighting Iterate = 1`. Setting this value to `0` will bring a preview completely off. Specialist will have to dig in the code to find out what explain this, then we'll be able to find a solution (or not).
